core/src/main/java/com/linecorp/armeria/common/PublisherBasedHttpResponse.java

```diff
@@ -25,4 +30,86 @@
     PublisherBasedHttpResponse(Publisher<? extends HttpObject> publisher) {
         super(publisher);
     }
+
+    static PublisherBasedHttpResponse from(ResponseHeaders headers,
+                                           Publisher<? extends HttpData> contentPublisher) {
+        return new PublisherBasedHttpResponse(new HeadersAndContentProcessor(headers, contentPublisher));
+    }
+
+    static final class HeadersAndContentProcessor implements Processor<HttpData, HttpObject> {
```

Comment:
- How about generalizing this class such as `PrependingPublisher`? (could be in `internal.common.stream`)

  ```java
  class PrependingPublisher<T> implements Publisher<T> {
 
      PrependingPublisher(T first, Publisher<? extends T> publisher) {
          ...
      }
  }
  ```
- It would be better to replace `Processor` with `Publisher` if we generalize it.
- The `Subscriber` interface could move to `HeadersAndContentSubscription` (perhaps need to rename it
  - By moving `Subscriber` to `SubscriptionImpl`, we don't need to care if `subscriber` is null
    ```java
    class SubscriptionImpl implements Subscriber<T>, Subscription {

        // This is non-null always.
	    private final Subscriber<T> subscriber;

        SubscriptionImpl (Subscriber<T> subscriber) {
        
       }
    }
    ```
- How about adding Reactive Streams [TCK](https://github.com/reactive-streams/reactive-streams-jvm/tree/master/tck) for the `PrependingPublisher`.

Reply:
Ah yes it is indeed cleaner. Here's what I've put together. Let me know what you think.

<details><summary>Code</summary>

```java
static class PrependingPublisher<T> implements Publisher<T> {

    private final T first;
    private final Publisher<? extends T> rest;

    PrependingPublisher(T first, Publisher<? extends T> rest) {
        this.first = first;
        this.rest = rest;
    }

    @Override
    public void subscribe(Subscriber<? super T> subscriber) {
        final RestSubscriber restSubscriber = new RestSubscriber(subscriber);
        rest.subscribe(restSubscriber);
    }

    class RestSubscriber implements Subscriber<T>, Subscription {

        private final Subscriber<? super T> subscriber;
        @Nullable
        private volatile Subscription subscription;
        @Nullable
        private volatile Throwable error;
        private volatile boolean completed;
        private volatile boolean firstSent;

        RestSubscriber(Subscriber<? super T> subscriber) {
            this.subscriber = subscriber;
        }

        @Override
        public void onSubscribe(Subscription subscription) {
            this.subscription = subscription;
            subscriber.onSubscribe(this);
        }

        @Override
        public void onNext(T t) {
            subscriber.onNext(t);
        }

        @Override
        public void onError(Throwable t) {
            // delay onError until the first piece is sent
            if (!firstSent) {
                error = t;
            } else {
                subscriber.onError(t);
            }
        }

        @Override
        public void onComplete() {
            // delay onComplete until the first piece is sent
            if (!firstSent) {
                completed = true;
            } else {
                subscriber.onComplete();
            }
        }

        @Override
        public void request(long n) {
            if (!firstSent) {
                subscriber.onNext(first);
                n--;
                firstSent = true;
            }
            if (n > 0) {
                if (error != null) {
                    subscriber.onError(error);
                } else if (completed) {
                    subscriber.onComplete();
                } else {
                    subscription.request(n);
                }
            }
        }

        @Override
        public void cancel() {
            subscription.cancel();
        }
    }
}
```

</details>

Reply:
Yeah. It looks nice to me. 👍

Reply:
I'll look into the TCK. Seems like many things are failing now 🤣 Is it the goal to pass all TCK cases?

Reply:
I've added the TCK tests. There were exceptions for rules 1.4 and 1.9 which I had to override some code (I've put comments there). Please take a look!
https://github.com/line/armeria/blob/afd7938b14681af8941a1cd450d25b9904372d56/core/src/test/java/com/linecorp/armeria/internal/common/stream/PrependingPublisherTckTest.java#L47

Reply:
> Is it the goal to pass all TCK cases?

Yes. If it is possible. 😉 However, we can ignore it in certain cases.